### PR TITLE
Add a config option to whitelist common initializer values for liveness analysis

### PR DESCRIPTION
**Pull request by jobnz**
Implements #1083

### CODE REVIEW

File: infer/src/checkers/liveness.ml [hunk 1]
```diff
@@ -219,11 +219,23 @@ let checker {IntraproceduralAnalysis.proc_desc; err_log} =
...
      use dead stores to a "sentinel" value as a heuristic for ignoring this case *)
+  let is_whitelisted = function
+    | Some c ->
+        let rec helper c = function
+          | hd :: lst ->
+              if Int.(hd = c) then true else helper c lst
+          | [] ->
+              false
+        in
+        helper c Config.liveness_whitelist_constant
```
Comment (jvillard) [inline infer/src/checkers/liveness.ml:230]: This looks like `List.mem ~equal:Int.equal Config.liveness_whitelist_constant c`.

File: infer/src/checkers/liveness.ml [hunk 2]
```diff
@@ -219,11 +219,23 @@ let checker {IntraproceduralAnalysis.proc_desc; err_log} =
...
+    | None ->
+        false
+  in
   let rec is_sentinel_exp = function
     | Exp.Cast (_, e) ->
         is_sentinel_exp e
     | Exp.Const (Cint i) ->
-        IntLit.iszero i || IntLit.isnull i
+        IntLit.iszero i || IntLit.isnull i || is_whitelisted (IntLit.to_int i)
```
Comment (jvillard) [inline infer/src/checkers/liveness.ml:238]: I could imagine someone wanting to whitelist, eg MAX_INT, which wouldn't fit in an OCaml `int` but would fit in an `IntLit.t`. To handle this, the config option could be just a list of strings that is pre-processed somewhere in this module (liveness.ml) into a list (or better yet, a set) of `IntLit.t`.

**Review by jvillard — COMMENTED** [review]
Looks useful to have, thank you! If you update your PR, could you also run `make manuals` to update the checked-in man pages?

File: infer/src/checkers/liveness.ml [hunk 3]
```diff
@@ -211,6 +211,13 @@ let get_captured_by_ref_invariant_map proc_desc =
   let cfg = ProcCfg.Exceptional.from_pdesc proc_desc in
   CapturedByRefAnalyzer.exec_cfg cfg () ~initial:VarSet.empty
 
+module IntLitSet = Caml.Set.Make (IntLit)
+
+let whitelisted_constants =
+  let int_lit_constants =
+    List.map ~f:(fun el -> IntLit.of_string el) Config.liveness_whitelist_constant
```
Comment (jvillard) [inline infer/src/checkers/liveness.ml:218]: If the user passes bad input we'll crash here. Maybe worth getting a nice error message? E.g.

```
fun el -> try IntLit.of_string el with Invalid_argument -> L.die UserError "Ill-formed option  '%s' for --liveness-whitelist-constant: an integer was expected"
```

**Review by jvillard — COMMENTED** [review]
Great stuff, thank you!

File: infer/src/checkers/liveness.ml (hunk 3 above)
Comment (jobnz) [reply infer/src/checkers/liveness.ml:218]: Good point. Thank you for the helpful feedback!

**Comment by jobnz** [comment]
Sorry, I'll undo the rebase

**Comment by jvillard** [comment]
Thanks a lot @jobnz! I took the liberty of doing some minor changes on top of your PR and adding a test (see 606a3c95d9d9ca4b9ae7015b9c991c04e4d20ed3), hope that's ok with you.

**Comment by jobnz** [comment]
Of course. Thanks again for the helpful reviews and comments!

**Pull request closed without merging**